scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,361 @@
+                                   stderr=subprocess.PIPE)
+        # `subprocess.communicate` returns a byte stream
+        stdout, stderr = process.communicate()
+        stdout, stderr = stdout.decode(), stderr.decode()
+
+    if process.returncode != 0:
+        msg = ("Something went wrong while deleting the "
+               "export {0:s}: stdout={1:s}, "
+               "stderr={2:s}".format(export_name, stdout, stderr))
+        if "SBC_ERR_NO_SUCH_SERVICE" in stderr:
+            log.warning(msg)
+        else:
+            raise EnvironmentError(msg)
+
+
+def _set_hosts_allow(export_name, hosts_allow):
+    """
+    Set the `hosts_allow` parameter for a given share.
+
+    :param export_name: name of export to grant access to
+    :type export_name: string (unicode)
+    :param hosts_allow: hosts allowed on this share
+    :type hosts_allow: iterable of `str`
+    """
+
+    cmd = ['net', 'conf', 'setparm', export_name,
+           'hosts allow', ' '.join(hosts_allow)]
+    process = subprocess.Popen(cmd, stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE)
+    # `subprocess.communicate` returns a byte stream
+    stdout, stderr = process.communicate()
+    stdout, stderr = stdout.decode(), stderr.decode()
+
+    if process.returncode != 0:
+        msg = ("Something went wrong while setting '{0!r}' as "
+               "the list of 'hosts allow' for share '{1:s}': stdout={2:s}, "
+               "stderr={3:s}".format(hosts_allow, export_name, stdout, stderr))
+        raise EnvironmentError(msg)
+
+
+@ensure_environment
+@ensure_export_exists
+def grant_access(export_name, host, exports, *args, **kwargs):
+    """
+    Grant access for a host to an export.
+
+    :param export_name: name of export to grant access to
+    :type export_name: string (unicode)
+    :param host: host to grant access for
+    :type host: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    """
+
+    hosts_allow = exports[export_name]['hosts allow'].split()
+
+    if host in hosts_allow:
+        msg = "Host '{0:s}' already allowed on share '{1:s}'".format(
+            host, export_name)
+        raise ClientExistsException(msg)
+
+    hosts_allow.append(host)
+
+    with utils.elevated_privileges():
+        _set_hosts_allow(export_name, hosts_allow)
+
+
+@ensure_environment
+@ensure_export_exists
+def revoke_access(export_name, host, exports, *args, **kwargs):
+    """
+    Revoke access for a host to an export.
+
+    :param export_name: name of export for revocation
+    :type export_name: string (unicode)
+    :param host: host to revoke access for
+    :type host: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    """
+
+    hosts_allow = exports[export_name]['hosts allow'].split()
+
+    try:
+        hosts_allow.remove(host)
+    except ValueError:
```

Comment:
Not very fond of the 'just go and handle errors later' pattern here. `hosts_allow` is a local, non-volatile variable, so doing a check first, raising appropriately and then taking action makes intent more clear.